deps/spidershim/src/v8functiontemplate.cc

```diff
@@ -0,0 +1,315 @@
+                                              Handle<Signature> signature,
+                                              int length)
+{
+  Local<Template> templ = Template::New(isolate, &functionTemplateClass);
+  if (*templ) {
+    return Local<FunctionTemplate>();
+  }
+
+  JSContext* cx = JSContextFromIsolate(isolate);
+  JS::RootedObject obj(cx, GetObject(*templ));
+  assert(obj);
+
+  SetCallbackAndData(obj, callback, data);
+
+  // TODO: figure out what should happen with signature.  What is it supposed to
+  // do?
+  js::SetReservedSlot(obj, LengthSlot, JS::Int32Value(length));
+
+  return Local<FunctionTemplate>(static_cast<FunctionTemplate*>(*templ));
+}
+
+Local<Function> FunctionTemplate::GetFunction()
+{
+  MaybeLocal<Function> maybeFunc =
+    GetFunction(Isolate::GetCurrent()->GetCurrentContext());
+  return maybeFunc.FromMaybe(Local<Function>());
+}
+
+MaybeLocal<Function> FunctionTemplate::GetFunction(Local<Context> context)
+{
+  JSContext* cx = JSContextFromContext(*context);
+  JS::RootedObject obj(cx, GetObject(this));
+  assert(obj);
+
+  // XXXbz V8 uses a temporary HandleScope here, but I _think_ that's some sort
+  // of performance optimization...
+
+  if (TemplateIsInstantiated(obj)) {
+    JS::RootedValue existingInstance(cx,
+                                     js::GetReservedSlot(obj, InstanceSlot));
+    assert(existingInstance.isObject());
+    return internal::Local<Function>::New(context->GetIsolate(),
+                                          existingInstance);
+  }
+
+  JS::RootedValue callbackVal(cx, js::GetReservedSlot(obj, CallbackSlot));
+  // XXXbz the .get() here is working around
+  // https://bugzilla.mozilla.org/show_bug.cgi?id=1275082
+  auto callback = reinterpret_cast<FunctionCallback>(callbackVal.get().toPrivate());
+
+  JS::RootedValue dataVal(cx, js::GetReservedSlot(obj, DataSlot));
+  // XXXbz is this a sane thing to do?  Or do I need to register this thing with
+  // some HandleScope?  Seems like I need to do the latter if the callee has
+  // this thing outlive the Rooted, right?
+  Local<Value> dataV8Val = GetV8Value(dataVal.address());
+
+  JS::RootedValue lengthVal(cx, js::GetReservedSlot(obj, LengthSlot));
+
+  // Figure out our prototype's prototype object.
+  JS::RootedObject protoProto(cx);
+  JS::RootedValue parentVal(cx, js::GetReservedSlot(obj, ParentSlot));
+  if (!parentVal.isUndefined()) {
+    assert(parentVal.isObject());
+    // XXXbz this seems a bit fishy... is it?
+    auto parentTemplate =
+      static_cast<FunctionTemplate*>(GetV8Template(parentVal.address()));
+
+    MaybeLocal<Function> parentFunc = parentTemplate->GetFunction(context);
+    if (parentFunc.IsEmpty()) {
+      return MaybeLocal<Function>();
+    }
+
+    JS::RootedObject parentObj(cx, GetObject(parentFunc.ToLocalChecked()));
+    assert(parentObj);
+
+    JS::RootedValue protoProtoVal(cx);
+    if (!JS_GetProperty(cx, parentObj, "prototype", &protoProtoVal)) {
```

Comment:
No, definitely not.  I'm getting function.prototype, not function.__proto__.  This is the object that's supposed to be used as the proto for instance objects created when using this function as a ctor.